src/TokenBuyer.sol

```diff
@@ -0,0 +1,377 @@
+     * by calling the callback function IBuyETHCallback#buyETHCallback.
+     * @param tokenAmount the amount of ERC20 tokens msg.sender wishes to sell to this contract in exchange for ETH, in token decimals.
+     * @param to the address to send ETH to by calling the callback function on it
+     * @param data arbitrary data passed through by the caller, usually used for callback verification
+     */
+    function buyETH(
+        uint256 tokenAmount,
+        address to,
+        bytes calldata data
+    ) external nonReentrant whenNotPaused {
+        uint256 amount = Math.min(tokenAmount, tokenAmountNeeded());
+        address _payer = payer;
+        uint256 balanceBefore = paymentToken.balanceOf(_payer);
+        uint256 ethAmount = ethAmountPerTokenAmount(amount);
+
+        IBuyETHCallback(to).buyETHCallback{ value: ethAmount }(msg.sender, amount, data);
+
+        uint256 tokensReceived = paymentToken.balanceOf(_payer) - balanceBefore;
+        if (tokensReceived < amount) {
+            revert ReceivedInsufficientTokens(amount, tokensReceived);
+        }
+
+        emit SoldETH(to, ethAmount, amount);
+    }
+
+    /**
+     * @notice Allow ETH top-ups.
+     */
+    receive() external payable {}
+
+    /**
+     ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
+      VIEW FUNCTIONS
+     ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
+     */
+
+    /**
+     * @notice Get how much ETH this contract needs in order to fund its current obligations plus `additionalTokens`, with
+     * a safety buffer `bufferBPs` basis points.
+     * @param additionalTokens an additional amount of `paymentToken` liability to use in this ETH requirement calculation, in payment token decimals.
+     * @param bufferBPs the number of basis points to add on top of the token liability price in ETH as a safety buffer, e.g.
+     * if `bufferBPs` is 10K, the function will return twice the amount it needs according to price alone.
+     * @return uint256 the amount of ETH needed to fund `additionalTokens` with a `bufferBPs` safety buffer.
+     */
+    function ethNeeded(uint256 additionalTokens, uint256 bufferBPs) public view returns (uint256) {
+        uint256 tokenAmount = tokenAmountNeeded() + additionalTokens;
+        uint256 ethCostOfTokens = ethAmountPerTokenAmount(tokenAmount);
+        uint256 ethCostWithBuffer = (ethCostOfTokens * (bufferBPs + 10_000)) / 10_000;
+
+        return ethCostWithBuffer - address(this).balance;
+    }
+
+    function tokenAmountNeededAndETHPayout() public view returns (uint256 tokenAmount, uint256 ethAmount) {
+        tokenAmount = tokenAmountNeeded();
+        ethAmount = ethAmountPerTokenAmount(tokenAmount);
+    }
+
+    /**
+     * @return uint256 the amount of `paymentToken` this contract is willing to buy in exchange for ETH, in payment token decimals.
+     */
+    function tokenAmountNeeded() public view returns (uint256) {
+        uint256 _paymentTokenBalance = paymentToken.balanceOf(address(this));
```

Comment:
@eladmallel this should be `paymentToken.balanceOf(payer)`, no?
if so, let's make sure we write a good test to catch this, e,g:
buyETH called once and fills USDC
buyETH called again buy no USDC needed any more

Reply:
there were tests for this, they were just buggy :) fixed the tests that turned red, then fixed this